R/foreign-keys.R

```diff
@@ -185,23 +188,32 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
   if (is_empty(ref_tbl_pk)) {
     abort_ref_tbl_has_no_pk(ref_table_name)
   }
-
-  tbl <- cdm_get_tables(dm)[[table_name]]
+  ref_tbl <- tbl(dm, ref_table_name)
+  tbl <- tbl(dm, table_name)
   tbl_colnames <- colnames(tbl)
 
-  ref_tbl <- cdm_get_tables(dm)[[ref_table_name]]
-
-  subsets <- map_lgl(
-    tbl_colnames,
-    ~ is_subset(tbl, !!.x, ref_tbl, !!ref_tbl_pk)
-  )
-
   tibble(
-    ref_table = ref_table_name,
-    ref_table_pk = ref_tbl_pk,
-    table = table_name,
     column = tbl_colnames,
-    candidate = subsets,
-    why = if_else(subsets, "", paste0("not a subset of ", ref_table, "$", ref_table_pk))
-  )
+    why = map_chr(column, ~why(dm, tbl, .x, ref_tbl, ref_table_name, ref_tbl_pk))
+  ) %>%
+    mutate(candidate = ifelse(why == "", TRUE, FALSE)) %>%
+    select(column, candidate, why) %>%
+    mutate(arrange_col = str_sub(why, 1, 3)) %>%
+    arrange(desc(candidate), desc(arrange_col), column) %>%
+    select(-arrange_col)
 })
+
+why <- function(dm, t1, colname, t2, t2_name, pk) {
+  names(pk) <- colname
+  test <- tryCatch(anti_join(
+    select(t1, !!sym(colname)), select(t2, !!sym(pk)), by = pk) %>%
```

Comment:
~~We need to remove dupes on the server: anti_join %>% distinct %>% arrange %>% head %>% collect %>% pull~~

We also might want the total count and the mismatch percentage to detect weak relationships:

```r
t1_join <- t1 %>% select(value = !!sym(colname))
t2_join <- t2 %>% select(value = !!sym(pk)) %>% mutate(match = 1L)

left_join(t1_join, t2_join, by = "value") %>%
  mutate(mismatch_or_null = if_else(is.na(match), value, NULL)) %>%
  count(mismatch_or_null) %>%
  ... %>%
  collect()
```

Example result for `flights$dest`:

1334 rows (8.5%) of the values not in `airports$faa`: SJU (280), BQN (110), 3 more, …